IDE/STM32Cube/wolfssl_example.c

```diff
@@ -1468,25 +1471,278 @@ static void PrintMemStats(void)
+
+typedef struct {
+    int curr_index;
+    int data_len;
+    char buf[MAX_RECORD_SIZE];
+} tls13_buf;
+
+/* This sets which UART to do the TLS 1.3 connection over.  It is something you
+ * will have to configure in STMCubeIDE and then change here. */
+#ifndef TLS_UART
+#define TLS_UART huart2
+#endif
+extern UART_HandleTypeDef TLS_UART;
+
+static int msg_length = 0;
+
+void HAL_UARTEx_RxEventCallback(UART_HandleTypeDef *huart, uint16_t Size) {
+   if (huart->Instance == TLS_UART.Instance) {
+       msg_length = Size;
+   }
+}
+
+static int uartIORx(WOLFSSL *ssl, char *buf, int sz, void *ctx)
 {
-	void* ptr = malloc(size);
-	if (ptr == NULL) {
-		printf("BREAK!\n");
-	}
-	return ptr;
+    HAL_StatusTypeDef status;
+    tls13_buf *tb = ctx;
+
+#ifdef DEBUG_UART_IO
+    printf("UART Read: In %d\n", sz);
+#endif
+
+    if (tb->curr_index + sz <= tb->data_len) {
+        XMEMCPY(buf, tb->buf + tb->curr_index, sz);
+        tb->curr_index += sz;
+#ifdef DEBUG_UART_IO
+        printf("UART Read1: Out %d\n", sz);
+#endif
+        return sz;
+    }
+
+    msg_length = 0;
+    XMEMSET(tb, 0, sizeof(*tb));
+
+    /* Now get the bytes sent */
+    status = HAL_UARTEx_ReceiveToIdle_DMA(&TLS_UART, (uint8_t *)tb->buf, MAX_RECORD_SIZE);
+
+    if (status != HAL_OK) {
+        return WOLFSSL_CBIO_ERR_WANT_READ;
+    } else {
+        /* We now go into an infinite loop waiting for msg_length to be set to a
+         * value other than 0. This will be done when the other side writes to
+         * UART and then idles. That will trigger HAL_UARTEx_RxEventCallback()
+         * which will seetee msg_length to the length of data written.
+         *
+         * If you mistakenly get stuck here, please simply reset the board.
+         */
+        while(msg_length == 0) {
+            HAL_Delay(10);
+        }
+#ifdef DEBUG_UART_IO
+        printf("Message received! length = %d\n", msg_length);
+#endif
+    }
+
+    /* now return the number of bytes requested. */
+    XMEMCPY(buf, tb->buf, sz);
+    tb->data_len = msg_length;
+    tb->curr_index = sz;
+
+#ifdef DEBUG_UART_IO
+    printf("UART Read2: Out %d\n", tb->data_len);
+#endif
+
+    return sz;
 }
-static void wolfFreeCb(void *ptr)
+
+static int uartIOTx(WOLFSSL *ssl, char *buf, int sz, void *ctx)
 {
-	free(ptr);
+    HAL_StatusTypeDef status;
+    int ret = sz;
+    (void)ctx;
+
+#ifdef DEBUG_UART_IO
+    printf("UART Write: In %d\n", sz);
+#endif
+
+    status = HAL_UART_Transmit(&TLS_UART, (uint8_t *)buf, sz, 0xFFFF);
+    if (status != HAL_OK) {
+        ret = WOLFSSL_CBIO_ERR_WANT_WRITE;
+    }
+
+#ifdef DEBUG_UART_IO
+    printf("UART Write: Out %d\n", ret);
+#endif
+
+    return ret;
 }
-static void* wolfReallocCb(void *ptr, size_t size)
+
+/* UART TLS 1.3 client and server */
+#ifndef NO_WOLFSSL_SERVER
+static int tls13_uart_server(void)
 {
-	return realloc(ptr, size);
+    int ret = -1, err;
+    WOLFSSL_CTX* ctx = NULL;
+    WOLFSSL* ssl = NULL;
+    byte echoBuffer[100];
+    tls13_buf tbuf;
```

Comment:
This is a huge stack buffer.... might be nice to have something like `WOLFSSL_SMALL_STACK` to allocate from the heap if needed.

Reply:
got it!